daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala

```diff
@@ -235,13 +335,349 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
+               x: Unit <- exercise @M:T Archive cId sig
+             in fetch @M:T cId""",
+            Array(SParty(alice), SParty(bob)),
+            alice,
+          )
+
+          inside(res) { case Success(Left(SErrorDamlException(IE.ContractNotActive(_, T, _)))) => }
+          msgs shouldBe createTraces :+ "archive"
+        }
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of fetch of an wrongly typed local contract
+        "wrongly typed contract" in {
+
+          val (res, msgs) = evalUpdateApp(
+            pkgs,
+            e"""\(sig : Party) ->
+             ubind cId1: ContractId M:Dummy <- create @M:Dummy M:Dummy { signatory = sig } 
+             in let cId2: ContractId M:T = COERCE_CONTRACT_ID @M:Dummy @M:T cId1
+             in fetch @M:T cId2""",
+            Array(SParty(alice)),
+            alice,
+          )
+
+          inside(res) {
+            case Success(Left(SErrorDamlException(IE.WronglyTypedContract(_, T, Dummy)))) =>
+          }
+          msgs shouldBe Seq.empty
+        }
+      }
+
+      // TEST_EVIDENCE: Semantics: Evaluation order of fetch of an unknown contract
+      "unknown contract" in {
+
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(cId: ContractId M:T) -> fetch @M:T cId""",
+          Array(SContractId(cId)),
+          alice,
+          getContract = PartialFunction.empty,
+        )
+
+        inside(res) { case Failure(SpeedyTestLib.UnknownContract(`cId`)) => }
+        msgs shouldBe localFetchByIdTraces
+      }
+
+    }
+
+    "fetch_by_key" - {
+
+      "a global contract" - {
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of successful fetch_by_key of a global contract
+        "success" in {
 
+          val (res, msgs) = evalUpdateApp(
+            pkgs,
+            e"""\(sig: Party) -> fetch_by_key @M:T (M:toKey sig)""",
+            Array(SParty(alice)),
+            alice,
+            getContract = getContract,
+            getKey = getKey,
+          )
+
+          inside(res) { case Success(Right(_)) => }
+          msgs shouldBe globalFetchByKeyTraces
+        }
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of fetch_by_key of an inactive global contract
+        "inactive contract" in {
+
+          val (res, msgs) = evalUpdateApp(
+            pkgs,
+            e"""\(cId: ContractId M:T) (sig: Party) -> 
+             ubind x: Unit <- exercise @M:T Archive cId sig
+             in fetch_by_key @M:T (M:toKey sig)""",
+            Array(SContractId(cId), SParty(alice)),
+            alice,
+            getContract = getContract,
+            getKey = getKey,
+          )
+
+          inside(res) { case Success(Left(SErrorDamlException(IE.ContractKeyNotFound(key)))) =>
+            key.templateId shouldBe T
+          }
+          msgs shouldBe globalFetchByIdTraces :+ "archive" :++ localFetchByKeyTraces
+        }
+
```

Comment:
I assume it is not possible to fetch_by_key a wrongly type contract.

Reply:
I think that’s right if we assume the ledger behaves correctly but maybe still worth checking that it throws an internal error?